clusterctl/cmd/validate_cluster.go

```diff
@@ -17,26 +17,130 @@ limitations under the License.
+		&vco.Namespace, "namespace", "n", v1.NamespaceDefault,
+		"If present, validate the cluster in the specified namespace.")
 	validateCmd.AddCommand(validateClusterCmd)
 }
 
-func RunValidateCluster() error {
-	return errors.NotImplementedError
+func RunValidateCluster() {
+	// TODO(wangzhen127): Also validate the cluster in addition to the cluster API objects.
+	cluster := validateClusterAPIObjects()
+
+	glog.Infof("Pass. The cluster %q in namespace %q is validated.", cluster.Name, vco.Namespace)
+}
+
+func validateClusterAPIObjects() *v1alpha1.Cluster {
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		clientcmd.NewDefaultClientConfigLoadingRules(),
+		&clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		glog.Fatalf("Failed to create client for talking to the apiserver: %v", err)
+	}
+
+	client, err := clientset.NewForConfig(config)
+	if err != nil {
+		glog.Fatalf("Failed to create client for talking to the apiserver: %v", err)
+	}
+
+	cluster, err := getClusterObject(client, vco.ClusterName, vco.Namespace)
+	if err != nil {
+		glog.Fatal(err)
+	}
+	err = validateClusterObject(cluster)
+	if err != nil {
+		glog.Fatal(err)
+	}
+	machines, err := client.ClusterV1alpha1().Machines(vco.Namespace).List(meta_v1.ListOptions{})
+	if err != nil {
+		glog.Fatalf("Failed to get the machines from the apiserver in namespace %q: %v", vco.Namespace, err)
+	}
+	err = validateMachineObjects(machines)
+	if err != nil {
+		glog.Fatal(err)
+	}
+	return cluster
+}
+
+func getClusterObject(client *clientset.Clientset, clusterName string, namespace string) (*v1alpha1.Cluster, error) {
+	if clusterName != "" {
+		cluster, err := client.ClusterV1alpha1().Clusters(namespace).Get(clusterName, meta_v1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get the cluster %q from the apiserver in namespace %q: %v", clusterName, namespace, err)
+		}
+		return cluster, nil
+	}
+
+	clusters, err := client.ClusterV1alpha1().Clusters(namespace).List(meta_v1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get the clusters from the apiserver in namespace %q: %v", namespace, err)
+	}
+	if numOfClusters := len(clusters.Items); numOfClusters == 0 {
+		return nil, fmt.Errorf("Fail: No cluster exists in namespace %q.", namespace)
+	} else if numOfClusters > 1 {
+		return nil, fmt.Errorf("Fail: There are more than one cluster in namespace %q. Please specify --cluster-name.", namespace)
+	}
+	return &clusters.Items[0], nil
+}
+
+func validateClusterObject(cluster *v1alpha1.Cluster) error {
+	glog.Infof("Checking cluster object %q...", cluster.Name)
+	if cluster.Status.ErrorReason != "" || cluster.Status.ErrorMessage != "" {
+		return fmt.Errorf("Failed to validate cluster object %q (%v): %s",
+			cluster.Name, cluster.Status.ErrorReason, cluster.Status.ErrorMessage)
+	}
+	return nil
+}
+
+func validateMachineObjects(machines *v1alpha1.MachineList) error {
+	for _, machine := range machines.Items {
```

Comment:
Rather than report the first machine that has something wrong with it, it may be better if an error was returned / printed for each non-valid machine. 

Not entirely sure how this command is intended to be used, but if I wanted to use it to 'fix' my cluster, as a user I will need to run this, correct what is wrong with the first machine, run it again, etc.

Reply:
Good point. Updated.